src/compiler/transformers/classFields.ts

```diff
@@ -898,21 +1009,57 @@ namespace ts {
             return pendingExpressions || (pendingExpressions = []);
         }
 
-        function addPrivateIdentifierToEnvironment(name: PrivateIdentifier) {
-            const text = getTextOfPropertyName(name) as string;
-            const weakMapName = factory.createUniqueName("_" + text.substring(1), GeneratedIdentifierFlags.Optimistic | GeneratedIdentifierFlags.ReservedInNestedScopes);
-            hoistVariableDeclaration(weakMapName);
-            getPrivateIdentifierEnvironment().set(name.escapedText, { placement: PrivateIdentifierPlacement.InstanceField, weakMapName });
-            getPendingExpressions().push(
-                factory.createAssignment(
-                    weakMapName,
+        function addPrivateIdentifierToEnvironment(node: PrivateClassElementDeclaration) {
+            const text = getTextOfPropertyName(node.name) as string;
+            const accessControlObject = createHoistedVariableForPrivateName(text);
+            let info: PrivateIdentifierInfo;
+            const assignmentExpressions: Expression[] = [];
+
+            if (isPropertyDeclaration(node)) {
+                info = {
+                    placement: PrivateIdentifierPlacement.InstanceField,
+                    weakMapName: accessControlObject
+                };
+
+                assignmentExpressions.push(factory.createAssignment(
+                    accessControlObject,
                     factory.createNewExpression(
                         factory.createIdentifier("WeakMap"),
                         /*typeArguments*/ undefined,
                         []
                     )
-                )
-            );
+                ));
+            }
+            else if (isMethodDeclaration(node)) {
+                const functionName = createHoistedVariableForPrivateName(text);
+
+                info = {
+                    placement: PrivateIdentifierPlacement.InstanceMethod,
+                    weakSetName: accessControlObject,
+                    functionName
+                };
+
+                assignmentExpressions.push(factory.createAssignment(
+                    accessControlObject,
+                    factory.createNewExpression(
+                        factory.createIdentifier("WeakSet"),
+                        /*typeArguments*/ undefined,
+                        []
+                    )
+                ));
+            }
+            else {
+                return;
+            }
+
+            getPrivateIdentifierEnvironment().set(node.name.escapedText, info);
+            getPendingExpressions().push(...assignmentExpressions);
+        }
+
+        function createHoistedVariableForPrivateName(privateName: string): Identifier {
+            const name = factory.createUniqueName("_" + privateName.substring(1), GeneratedIdentifierFlags.Optimistic);

```

Comment:
This used to be `GeneratedIdentifierFlags.Optimistic | GeneratedIdentifierFlags.ReservedInNestedScopes` but `ReservedInNestedScopes` seems to be doing the opposite of what I would expect it to do. It caused shadowing generated variables in the upper scope. I will need to clarify this on the official PR.

Reply:
Input: 
```ts
class Base {
    #x() { };
    constructor() {
        class Derived {
            #x() { };
            testBase(x: Base) {
                console.log(x.#x);
            }
        }
    }
}
```

Output (`GeneratedIdentifierFlags.Optimistic`):

```js
var __classPrivateMethodGet = (this && this.__classPrivateMethodGet) || function (receiver, accessCheck, fn) {
    if (!accessCheck.has(receiver)) {
        throw new TypeError("attempted to get private method on non-instance");
    }
    return fn;
};
var _x, _x_1;
class Base {
    constructor() {
        var _x_2, _x_3;
        _x.add(this);
        class Derived {
            constructor() {
                _x_2.add(this);
            }
            ;
            testBase(x) {
                console.log(__classPrivateMethodGet(x, _x_2, _x_3));
            }
        }
        _x_2 = new WeakSet(), _x_3 = function _x_3() { };
    }
    ;
}
_x = new WeakSet(), _x_1 = function _x_1() { };
```

Output (`GeneratedIdentifierFlags.Optimistic | GeneratedIdentifierFlags.ReservedInNestedScopes`):

```js
var __classPrivateMethodGet = (this && this.__classPrivateMethodGet) || function (receiver, accessCheck, fn) {
    if (!accessCheck.has(receiver)) {
        throw new TypeError("attempted to get private method on non-instance");
    }
    return fn;
};
var _x, _x_1;
class Base {
    constructor() {
        var _x_2, _x;
        _x.add(this);
        class Derived {
            constructor() {
                _x_2.add(this);
            }
            ;
            testBase(x) {
                console.log(__classPrivateMethodGet(x, _x_2, _x));
            }
        }
        _x_2 = new WeakSet(), _x = function _x() { };
    }
    ;
}
_x = new WeakSet(), _x_1 = function _x_1() { };
```

`_x_2` is correctly generated. But the next one reuses `_x` instead of `_x_3`, which is wrong.

Reply:
`ReservedInNestedScopes` seems to *restrict* the scope in which the name is reserved. (Names generated without that flag are guaranteed to not conflict with any other generated names.) I imagine you'd only want to use it for temporary variables with limited scope (say a loop counter for example, where you do not need it to be unique within the file). The name generation logic lives in emitter.ts, look for `reservedNamesStack` and `generatedNames`.